docs/library/machine.Counter.rst

```diff
@@ -0,0 +1,67 @@
+.. currentmodule:: machine
+.. _machine.Counter:
+
+class Counter-- Pulse Counter
+=============================
+
+This class provides access to hardware-supported pulse counting.
+
+It is currently provided for ports:
+
+  * :ref:`ESP32 <esp32_machine.Counter>`
+  * :ref:`MIMXRT <mimxrt_machine.Counter>`
+
+Minimal example usage::
+
+    from machine import Pin, Counter
+
+    counter = Counter(0, src=Pin(0, mode=Pin.INPUT))  # create Counter object and start to count input pulses
+    counter.init(filter_ns=1000)                      # switch source filtering on
+    value = counter.value(0)                          # get current Counter value, set counter to 0
+    counter.deinit()                                  # turn off the Counter
+
+    print(counter)                                    # show the Counter object properties
+
+Constructor
+-----------
+
+.. class:: Counter(id, src=None, \*, direction=1, filter_ns=0)
+
+      - *id*. Values of *id* depend on a particular port and its hardware.
+        Values 0, 1, etc. are commonly used to select hardware block #0, #1, etc.
+
+      - *src*. The Counter pulses input pin, which is usually a
+        :ref:`machine.Pin <machine.Pin>` object, but a port may allow other values,
+        like integers or strings, which designate a Pin in the *machine.Pin* class.
+        It may be omitted on ports that have a predefined pin for *id*-specified hardware block.
+        The keyword may be omitted, for example, Counter(0, Pin(0)).
+
+      - *direction*\=value. Specifying the direction of counting. The default value is 1. Suitable values are:
+
+        - if value == 0 or False: count down -1
+        - if value != 0 or True: count up +1
+        - a :ref:`machine.Pin <machine.Pin>` object. The level at that pin controls
```

Comment:
The level for the count direction may change with the hardware. MIMXRT hardware counts up with a low level and counts down with a high level. That behavior cannot be changed by software.

Reply:
Is there still an issue here?

Reply:
I'd probably make clear that the Pin object is a port-specific extension. Something like:

```
- *direction* specifies the direction to count. Values for this include the constants ``Counter.UP``
  and ``Counter.DOWN``, with ``Counter.UP`` the default. Ports may support additional values or
  objects, such as a :ref:`machine.Pin <machine.Pin>` object to control the direction externally.
```